timemachine/cpp/src/gpu_utils.cuh

```diff
@@ -21,8 +22,22 @@ templateCurandNormal(curandGenerator_t generator, double *outputPtr, size_t n, d
 inline void gpuAssert(cudaError_t code, const char *file, int line, bool abort = true) {
     if (code != cudaSuccess) {
         fprintf(stderr, "GPUassert: %s %s %d\n", cudaGetErrorString(code), file, line);
-        if (abort)
+        if (abort) {
+            // If the GPU is invalid or missing for some reason, raise an exception so we can handle that
+            // Error codes can be found here: https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1g3f51e3575c2178246db0a94a430e0038
+            switch (code) {
+            case cudaErrorInvalidDevice:
+            case cudaErrorInsufficientDriver:
+            case cudaErrorNoDevice:
+            case cudaErrorStartupFailure:
+            case cudaErrorInvalidPtx:
+            case cudaErrorUnsupportedPtxVersion:
+                throw InvalidHardware();
```

Comment:
Would it be useful to include the error code in the exception?

Reply:
Would the error string work as the thing to return? As the codes won't be accessible from Python (though we could expose them)

Reply:
Not sure, just seems like it could be useful if there was a way to tell from the Python exception (possibly by referencing the CUDA API docs) which of these conditions we hit.

Reply:
Added code along with the message as a string in https://github.com/proteneer/timemachine/pull/1104/commits/b0723ab266a82782e07ab114063769b08ce6e3e0

Didn't add the int on the exception as it isn't clear how to do that and don't think it would be of much utility.